libraries/AP_Proximity/AP_Proximity_LightWareSF40C.cpp

```diff
@@ -166,280 +197,288 @@ void AP_Proximity_LightWareSF40C::init_sectors()
-            _motor_direction = atoi(element_buf[0]);
-            success = true;
-            break;
+    case ParseState::MSG_ID:
+        msg.msgid = (MessageID)b;
+        msg.crc_expected = crc_xmodem_update(msg.crc_expected, b);
+        if (msg.payload_len > 1) {
+            msg.state = ParseState::PAYLOAD;
+        } else {
+            msg.state = ParseState::CRC_L;
+        }
+        msg.payload_recv = 0;
+        break;
+
+    case ParseState::PAYLOAD:
+        if (msg.payload_recv < (msg.payload_len - 1)) {
+            msg.payload[msg.payload_recv] = b;
+            msg.payload_recv++;
+            msg.crc_expected = crc_xmodem_update(msg.crc_expected, b);
+        }
+        if (msg.payload_recv >= (msg.payload_len - 1)) {
+            msg.state = ParseState::CRC_L;
+        }
+        break;
+
+    case ParseState::CRC_L:
+        msg.crc_low = b;
+        msg.state = ParseState::CRC_H;
+        break;
+
+    case ParseState::CRC_H:
+        msg.crc_high = b;
+        if (msg.crc_expected == UINT16_VALUE(msg.crc_high, msg.crc_low)) {
+            process_message();
+            last_reply_ms = AP_HAL::millis();
+        }
+        msg.state = ParseState::HEADER;
+        break;
+    }
+}
 
-        case RequestType_ForwardDirection:
-            _forward_direction = atoi(element_buf[0]);
-            success = true;
+// process the latest message held in the msg structure
+void AP_Proximity_LightWareSF40C::process_message()
+{
+    // process payload
+    switch (msg.msgid) {
+    case MessageID::TOKEN:
+        // copy token into sensor_state.token variable
+        if (msg.payload_recv == ARRAY_SIZE(sensor_state.token)) {
+            memcpy(sensor_state.token, msg.payload, ARRAY_SIZE(sensor_state.token));
+        }
+        break;
+    case MessageID::RESET:
+        // no need to do anything
+        break;
+    case MessageID::STREAM:
+        if (msg.payload_recv == sizeof(uint32_t)) {
+            sensor_state.streaming = (buff_to_uint32(msg.payload[0], msg.payload[1], msg.payload[2], msg.payload[3]) == 3);
+        }
+        break;
+    case MessageID::DISTANCE_OUTPUT: {
+        last_distance_received_ms = AP_HAL::millis();
+        const uint16_t point_total = buff_to_uint16(msg.payload[8], msg.payload[9]);
+        const uint16_t point_count = buff_to_uint16(msg.payload[10], msg.payload[11]);
+        const uint16_t point_start_index = buff_to_uint16(msg.payload[12], msg.payload[13]);
+        // sanity check point_total
+        if (point_total == 0) {
             break;
-
-        case RequestType_DistanceMeasurement:
-        {
-            float angle_deg = (float)atof(element_buf[0]);
-            float distance_m = (float)atof(element_buf[1]);
+        }
+        const float angle_inc_deg = (1.0f / point_total) * 360.0f;
+        const float angle_sign = (frontend.get_orientation(state.instance) == 1) ? -1.0f : 1.0f;
+        const float angle_correction = frontend.get_yaw_correction(state.instance);
```

Comment:
We should be checking that the correction in the packet is 0, otherwise we need to subtract it.

Reply:
I'm not I understand this but I think what I've done is consistent with the other backends.

Checking this has made me notice that we need a wrap_360 here though!

Reply:
oh, I think you're referring to the forward offset field.. I haven't tested that but I wonder if consuming that would have the effect of removing an offset provided by the user?  In any case, I think for now we can ignore this forward offset maintained in the sensor.